src/test/java/uk/nhs/hee/tis/trainee/credentials/service/CredemtoalEvemtServiceTest.java

```diff
@@ -0,0 +1,176 @@
+ * DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+package uk.nhs.hee.tis.trainee.credentials.service;
+
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.UUID;
+
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.mockito.Mock;
+import uk.nhs.hee.tis.trainee.credentials.TestJwtUtil;
+import uk.nhs.hee.tis.trainee.credentials.dto.CredentialType;
+import uk.nhs.hee.tis.trainee.credentials.model.CredentialMetadata;
+import uk.nhs.hee.tis.trainee.credentials.repository.CredentialMetadataRepository;
+
+import static org.junit.Assert.assertEquals;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.times;
+
+class CredentialEventServiceTest {
+
+
+  private static final UUID PLACEMENT_1_credentialId =
+      UUID.fromString("e2a26a50-2c51-4327-8692-1cb05b80af07");
+  private static final String PLACEMENT_1_traineeId = "47165";
+  private static final String PLACEMENT_1_tisId = "2bd3f009-9c7a-4352-bcc6-9dff8eac07d6";
+  private static final Instant PLACEMENT_1_issuedAt =
+      Instant.parse("2023-09-27T09:37:48.000+00:00");
+
+  private static final UUID PLACEMENT_2_credentialId =
+      UUID.fromString("e5c09a44-39ff-442f-b356-8424caa581d0");
+  private static final String PLACEMENT_2_traineeId = "47165";
+  private static final String PLACEMENT_2_tisId = "2bd3f009-9c7a-4352-bcc6-9dff8eac07d6";
+  private static final Instant PLACEMENT_2_issuedAt =
+      Instant.parse("2023-08-14T14:12:54.000+00:00");
+
+  private static final UUID PLACEMENT_3_credentialId =
+      UUID.fromString("6ada9134-a582-44a2-8ce0-106ae8b9bcd4");
+  private static final String PLACEMENT_3_traineeId = "47165";
+  private static final String PLACEMENT_3_tisId =
+      "72055636-1439-11ee-be56-0242ac120002";
+  private static final Instant PLACEMENT_3_issuedAt =
+      Instant.parse("2023-09-27T09:41:35.000+00:00");
+
+  private static final UUID PROGRAMME_1_credentialId =
+      UUID.fromString("6ada9134-a582-44a2-8ee0-125ae8b9bcd4");
+  private static final String PROGRAMME_1_traineeId = "47165";
+  private static final String PROGRAMME_1_tisId = "72055636-1439-11ee-be56-0242ac120002";
+  private static final Instant PROGRAMME_1_issuedAt =
+      Instant.parse("2023-09-27T09:41:35.000+00:00");
+
+  private static final UUID PROGRAMME_2_credentialId =
+      UUID.fromString("6ada9134-a582-44a2-8de0-105ae8b9bcd4");
+  private static final String PROGRAMME_2_traineeId = "47165";
+  private static final String PROGRAMME_2_tisId = "72055636-1439-11ee-be56-0242ac120002";
+  private static final Instant PROGRAMME_2_issuedAt =
+      Instant.parse("2023-08-27T09:41:35.000+00:00");
```

Comment:
It's hard to figure out what is and isn't meant to match here, you could initialise some of the values separately and reuse the same constant during `setUp()`.

e.g. refactoring as below makes it clear what the expectations are without having to mentally parse a load of hard coded IDs and timestamps. 


```java
private static final String TIS_ID_1 = UUID.randomUUID().toString();
private static final String TIS_ID_2 = UUID.randomUUID().toString();

private static final String TRAINEE_ID = "47165";

private static final String PLACEMENT_DISPLAY_NAME = CredentialType.TRAINING_PLACEMENT.getDisplayName();
private static final String PROGRAMME_DISPLAY_NAME = CredentialType.TRAINING_PROGRAMME.getDisplayName();

private CredentialMetadata placement1Cred1;
private CredentialMetadata placement1Cred2;
private CredentialMetadata placement2Cred1;

private CredentialMetadata programmeCred1;
private CredentialMetadata programmeCred2;

@BeforeEach
void setUp() {
  placement1Cred1 = new CredentialMetadata();
  placement1Cred1.setTisId(TIS_ID_1);
  placement1Cred1.setIssuedAt(Instant.MAX);
  placement1Cred1.setCredentialId(UUID.randomUUID().toString());
  placement1Cred1.setTraineeId(TRAINEE_ID);
  placement1Cred1.setCredentialType(PLACEMENT_DISPLAY_NAME);

  placement1Cred2 = new CredentialMetadata();
  placement1Cred2.setTisId(TIS_ID_1);
  placement1Cred2.setIssuedAt(Instant.MIN);
  placement1Cred2.setCredentialId(UUID.randomUUID().toString());
  placement1Cred2.setTraineeId(TRAINEE_ID);
  placement1Cred2.setCredentialType(PLACEMENT_DISPLAY_NAME);

  placement2Cred1 = new CredentialMetadata();
  placement2Cred1.setTisId(TIS_ID_2);
  placement2Cred1.setIssuedAt(Instant.MAX);
  placement2Cred1.setCredentialId(UUID.randomUUID().toString());
  placement2Cred1.setTraineeId(TRAINEE_ID);
  placement2Cred1.setCredentialType(PLACEMENT_DISPLAY_NAME);

  programmeCred1 = new CredentialMetadata();
  programmeCred1.setTisId(TIS_ID_1);
  programmeCred1.setIssuedAt(Instant.MAX);
  programmeCred1.setCredentialId(UUID.randomUUID().toString());
  programmeCred1.setTraineeId(TRAINEE_ID);
  programmeCred1.setCredentialType(PROGRAMME_DISPLAY_NAME);

  programmeCred2 = new CredentialMetadata();
  programmeCred2.setTisId(TIS_ID_1);
  programmeCred2.setIssuedAt(Instant.MIN);
  programmeCred2.setCredentialId(UUID.randomUUID().toString());
  programmeCred2.setTraineeId(TRAINEE_ID);
  programmeCred2.setCredentialType(PROGRAMME_DISPLAY_NAME);
}
```